brave-okhttp/src/main/java/com/github/kristofa/brave/okhttp/BraveTracingInterceptor.java

```diff
@@ -0,0 +1,238 @@
+     * network request. These span name of network requests are unaffected by this setting.
+     */
+    public Builder tagIsApplicationSpanName(boolean tagIsApplicationSpanName) {
+      this.tagIsApplicationSpanName = tagIsApplicationSpanName;
+      return this;
+    }
+
+    public BraveTracingInterceptor build() {
+      return new BraveTracingInterceptor(this);
+    }
+  }
+
+  final Brave brave;
+  final ClientRequestInterceptor requestInterceptor;
+  final ClientResponseInterceptor responseInterceptor;
+  final SpanNameProvider spanName;
+  final boolean tagIsApplicationSpanName;
+  final String serverName;
+
+  BraveTracingInterceptor(Builder builder) {
+    brave = builder.brave;
+    requestInterceptor = new ClientRequestInterceptor(brave.clientTracer());
+    responseInterceptor = new ClientResponseInterceptor(brave.clientTracer());
+    spanName = builder.spanNameProvider;
+    tagIsApplicationSpanName = builder.tagIsApplicationSpanName;
+    serverName = builder.serverName;
+  }
+
+  @Override
+  public Response intercept(Chain chain) throws IOException {
+    Request request = chain.request();
+    Connection connection = chain.connection();
+    if (connection == null) { // we are an application interceptor, so start a local span.
+      String name = tagIsApplicationSpanName && !(request.tag() instanceof Request)
+          ? request.tag().toString()
+          : spanName.spanName(new OkHttpRequest(request));
+      SpanId spanId = brave.localTracer().startNewSpan("okhttp", name);
+      if (spanId == null) request = request.newBuilder().addHeader(Sampled.getName(), "0").build();
+      try {
+        return chain.proceed(request);
+      } catch (IOException | RuntimeException | Error e) {
+        // TODO: revisit https://github.com/openzipkin/openzipkin.github.io/issues/52
+        brave.localTracer().submitBinaryAnnotation(Constants.ERROR, e.getMessage());
+        throw e;
+      } finally {
+        brave.localTracer().finishSpan(); // span must be closed!
+      }
+    } else { // start a client span for each network attempt
+      return traceRequest(chain, request, connection);
+    }
+  }
+
+  Response traceRequest(Chain chain, Request request, Connection connection) throws IOException {
+    OkHttpClientRequest okHttpRequest = new OkHttpClientRequest(request, connection);
+    try {
+      requestInterceptor.handle(new OkHttpClientRequestAdapter(okHttpRequest));
+      Response response = chain.proceed(okHttpRequest.builder.build());
+      responseInterceptor.handle(new HttpClientResponseAdapter(new OkHttpResponse(response)));
+      return response;
+    } catch (IOException | RuntimeException | Error e) {
+      // TODO: revisit https://github.com/openzipkin/openzipkin.github.io/issues/52
+      brave.clientTracer().submitAnnotation(Constants.ERROR);
+      brave.clientTracer().setClientReceived(); // span must be closed!
```

Comment:
probably should clarify the comment. responseInterceptor.handle internally calls setClientReceived last, so on throw we have to do it manually.